### ISSUE TITLE
CIFAR recover stage problems

### ISSUE THREAD

**Issue by j-cyoung** [shaoshitong/EDC#2]
I encountered several problems during the reproduction process(`data_synthesis_with_svd_with_db_with_all_statistic.py`):

1. lack of module name

```python
# started from line 54
    for i, (_model_teacher) in enumerate(model_teacher):
        print(_model_teacher)
        for name, module in _model_teacher.named_modules():
            if isinstance(module, nn.BatchNorm2d):
                loss_r_feature_layers[i].append(BNFeatureHook(module,training_momentum=args.training_momentum))
            elif isinstance(module, nn.Conv2d):
                _hook_module = ConvFeatureHook(module, save_path=args.statistic_path,
                                               name=str(_model_teacher.__class__.__name__) + "=" + name,
                                               gpu=gpu, training_momentum=args.training_momentum, drop_rate=args.drop_rate)
                _hook_module.set_hook(pre=True)
                load_tag = load_tag & _hook_module.load_tag
                loss_r_feature_layers[i].append(_hook_module)
        print(load_tag)
```

In the initialization of `BNFeatureHook`, it requires `name` attribute.

2. lack of targets name

In the `pre_hook_fn` and `post_hook_fn` of `ConvFeatureHook` class, the `self.targets` is needed, but the code in `data_synthesis_with_svd_with_db_with_all_statistic.py` didn't set this attribute. I check the `recover.py` file and it set `self.targets=targets` in the loop below. And in the code of `data_synthesis_with_svd_with_db_with_all_statistic.py`, this attribute wasn't set in the similar loop.

```python
# from file recover.py
# started from line 171
            for j, _model_teacher in enumerate(model_teacher):
                if not load_tag_dict[j]:
                    print(f"conduct backbone {args.aux_teacher[j]} statistics")
                    for i, (data, targets) in tqdm(enumerate(train_loader)):
                        data = data.cuda(gpu)
                        targets = targets.cuda(gpu)
                        for _loss_t_feature_layer in loss_r_feature_layers[j]:
                            _loss_t_feature_layer.set_label(targets)	# <- set self.targets
                        _ = _model_teacher(data)
                    
                    for _loss_t_feature_layer in loss_r_feature_layers[j]:
                        _loss_t_feature_layer.save()

# started from line 244
            for mod in loss_r_feature_layers[id]:
                mod.set_label(targets)	# <- set self.targets
```

3. incomplete aux_teacher

As said in the paper, the cifar's aux_teachers include `{ResNet-18, ConvNet-W128, MobileNet-V2, WRN-162, ShuffleNet-V2, ConvNet-D1, ConvNet-D2, ConvNet-W32}`. However, neither the code in `squeeze` dir nor `recover` dir include 8 models above. The scripts is mismatched with the paper and the code didn't prepare for all the 8 backbones too. 

```shell
# squeeze.sh
python train.py --model ResNet18 --dataset CIFAR-10 --data_path /data/spiderman/cifar10 --train_epochs 5

python train.py --model MobileNetV2 --dataset CIFAR-10 --data_path /path/to/cifar-10 --train_epochs 5

python train.py --model ShuffleNetV2_0_5 --dataset CIFAR-10 --data_path /path/to/cifar-10 --train_epochs 5

python train.py --model WRN_16_2 --dataset CIFAR-10 --data_path /path/to/cifar-10 --train_epochs 5

python train.py --model ConvNetW128 --dataset CIFAR-10 --data_path /path/to/cifar-10 --train_epochs 5
```

```python
# squeeze>train.py, from line 81
# it seems just 5 backbones
    if args.model in ["resnet18","mobilenet_v2","efficientnet_b0","shufflenet_v2_x0_5"]:
        model = models.__dict__[args.model](pretrained=True, num_classes=1000)
        if args.model == "resnet18":
            model.conv1 = nn.Conv2d(
                3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False
            )
            model.maxpool = nn.Identity()
            model.fc = nn.Linear(model.fc.in_features,10)
        elif args.model == "mobilenet_v2":
            model.features[0][0] = nn.Conv2d(
                3, model.features[0][0].out_channels, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False
            )
            model.classifier[1] = nn.Linear(model.classifier[1].in_features,10)
        elif args.model == "efficientnet_b0":
            model.features[0][0] = nn.Conv2d(
                3, model.features[0][0].out_channels, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False
            )
            model.classifier[1] = nn.Linear(model.classifier[1].in_features,10)
        elif args.model == "shufflenet_v2_x0_5":
            model.conv1 = nn.Conv2d(
                    3, model.conv1[0].out_channels, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False
                )
            model.maxpool = nn.Identity()
            model.fc = nn.Linear(model.fc.in_features,10)
    else:
        if args.model == "ConvNetW128":
            model = ti_get_network(args.model , channel = 3, num_classes = 10, im_size = (32, 32), dist = False)
        else:
            model = model_dict[args.model](num_classes = 10)
```

```python
# recover>data_synthesis_with_svd_with_db_with_all_statistic.py, from line 331
# only 5 backbones too
	aux_teacher = ["ResNet18", "ConvNetW128", "MobileNetV2", "WRN_16_2", "ShuffleNetV2_0_5"]
    model_teacher = []
    for name in aux_teacher:
        if name == "ConvNetW128":
            model = ti_get_network(name, channel=3, num_classes=10, im_size=(32, 32), dist=False)
        else:
            model = ti_models.model_dict[name](num_classes=10)

        model_teacher.append(model)
        checkpoint = torch.load(
            os.path.join(args.pre_train_path, "CIFAR-10", name, f"squeeze_{name}.pth"),
            map_location="cpu")
        model_teacher[-1].load_state_dict(checkpoint)
```

-----------------

All the bug I've tried to fix and made it run finally. However, I'm not sure whether I realize the code and fix it in a right way. So could you please fix these issues?

Thanks for your great work and sharing.

**Comment by ai23resch04001** [comment]
@j-cyoung Hi, It seems the issues are still not fixed in the main branch. 
Can you share the fixes that you have done (any fork you have made)?

Thank you.

**Comment by j-cyoung** [comment]
@ai23resch04001 I ran the wrong python script. Just run `recover.py` instead of `data_synthesis_with_svd_with_db_with_all_statistic.py`

**Comment by ai23resch04001** [comment]
@j-cyoung  Thanks for your response. Should we use 8 models or 5 models , for squeeze operation?

**Comment by j-cyoung** [comment]
@ai23resch04001 
I’m not entirely certain, but in my implementation, I trained 8 models. This decision was based on the observation that in the [CIFAR-10](https://github.com/shaoshitong/EDC/blob/79a0c3220a21bd1dfe764f1ad95f6eabe340bdeb/Branch_CIFAR_10/recover/recover.py#L430) and [CIFAR-100](https://github.com/shaoshitong/EDC/blob/79a0c3220a21bd1dfe764f1ad95f6eabe340bdeb/Branch_CIFAR_100/recover/recover.py#L384) recover.py scripts, 8 models are used. In contrast, the [ImageNet version](https://github.com/shaoshitong/EDC/blob/79a0c3220a21bd1dfe764f1ad95f6eabe340bdeb/Branch_full_ImageNet_1k/recover/recover.py#L386) utilizes only 5 models. It appears that using 8 models is a design choice specifically tailored for the CIFAR datasets.

Accordingly, I modified the squeeze script to train 8 models, and the results turned out to be good. You may consider experimenting with both configurations—using 5 and 8 models respectively—to determine which one yields better performance or aligns more closely with the results reported in the paper.

**Comment by ai23resch04001** [comment]
@j-cyoung Thanks for the info. 
Did you use the same hyperparameter setting as in "recover.sh" or did you use the ones mentioned in the github repo? 
(For example, recover.sh gives iterations=75 while the github repo command uses 4000)

**Comment by j-cyoung** [comment]
@ai23resch04001 I used 75 iterations, but for details about the specific hyperparameters, it might be best to consult the author directly.

**Comment by karan-kumawat17** [comment]
@j-cyoung Thank you for the info, I was struggling a bit with this.

I need a little clarification about few things:
1. So you trained 8 models right, if someone goes with 5 models in CIFAR100, will it create problem?
2. Can you tell the hyperparameters which you used in recover.py? As for the recover.sh, I am seeing iterations=2000, so needed to confirm on this too.
3. Did you use each architecture for your ablation study, I am interested in the results if you can show them.
4. Was there any problem in relabel.sh further after this?

**Comment by j-cyoung** [comment]
Hi @karan-kumawat17 , sorry for the delayed response.
	1.	Running with 5 models should work fine. The number of models is a hyperparameter—you just need to ensure that the models used in recover.py were pre-trained in squeeze.py.
	2.	Here’s the recover.sh script I used:

```shell
#!/bin/bash
data_path=/nas/dataset/dataset_distillation/cifar/cifar100
imgs_path=initial_imgs/500_0
rand_number=1

for ipc_number in 1 10
do
imgs_path=initial_imgs/500_$((rand_number-1))
CUDA_VISIBLE_DEVICES=0,1,2,3 python recover.py \
    --arch-name "resnet18" \
    --exp-name "EDC_CIFAR_100_Recover_IPC_${ipc_number}_${rand_number}" \
    --batch-size 100 --category-aware "global" \
    --lr 0.05 --drop-rate 0.0 \
    --ipc-number ${ipc_number} --training-momentum 0.8 \
    --iteration 2000 \
    --train-data-path  /nas/dataset/dataset_distillation/cifar/cifar100 \
    --r-loss 0.01 --initial-img-dir $imgs_path \
    --verifier --store-best-images --gpu-id 0,1,2,3 &> ./log/${ipc_number}_${rand_number} &
wait
done
```

Performance (IPC):
	•	IPC=1 → Accuracy: 42.99%
	•	IPC=10 → Accuracy: 65.34%

3.	I didn’t conduct an ablation study, so unfortunately I don’t have those results to share.
4.	Initializing the synthetic images in the recovery stage is critical. Since the authors didn’t release code for this, I wrote my own script to initialize the images by sampling from CIFAR-100. Here’s the script:

```python
import os
import argparse
import random
import torchvision.datasets as datasets

def initial_imgs_ori(original_path, save_path, IPC, seed):
    """Randomly select a subset from CIFAR-100 and save it"""

    # Set the random seed
    random.seed(seed)

    # Load CIFAR-100 training set
    cifar100_dataset = datasets.CIFAR100(root=original_path, train=True, download=True)
    
    # Group image indices by class
    class_indices = {i: [] for i in range(100)}
    for index, (_, label) in enumerate(cifar100_dataset):
        class_indices[label].append(index)

    # Create the output directory
    os.makedirs(save_path, exist_ok=True)

    # Randomly select IPC images per class and save them
    for class_id in range(100):
        class_images_indices = class_indices[class_id]
        selected_indices = random.sample(class_images_indices, IPC)

        # Create directory for each class
        class_folder = os.path.join(save_path, f"{class_id:03d}")
        os.makedirs(class_folder, exist_ok=True)

        # Save the selected images as PNGs
        for idx in selected_indices:
            img, _ = cifar100_dataset[idx]
            img.save(os.path.join(class_folder, f"{idx}.png"))

def main():
    parser = argparse.ArgumentParser(description='Randomly select a CIFAR-100 subset and save it')
    parser.add_argument('original_path', type=str, help='Path to the original CIFAR-100 dataset')
    parser.add_argument('save_path', type=str, help='Path to save the selected subset')
    parser.add_argument('IPC', type=int, help='Number of images per class to retain')
    parser.add_argument('--seed', type=int, default=42, help='Random seed')

    args = parser.parse_args()

    initial_imgs_ori(args.original_path, args.save_path, args.IPC, args.seed)

if __name__ == "__main__":
    main()
```

These codes are my own private implementations. For more authoritative or official clarifications, I recommend reaching out to the original authors @shaoshitong .

**Comment by ai23resch04001** [comment]
Hi @j-cyoung 
Thanks for this excellent explanation. 
Could you please provide the script you used for generating labels and training FKD models? 

Thank you.

**Comment by 2018cx** [comment]
> [@ai23resch04001](https://github.com/ai23resch04001) I ran the wrong python script. Just run `recover.py` instead of `data_synthesis_with_svd_with_db_with_all_statistic.py`
@j-cyoung Hi, how do you use recover.py to get the initial images?
